Create 2025-09-30-introduction-to-ai-agents-in-opensearch.md (#3961)
opensearch-project/project-website: _posts/2025-09-30-introduction-to-ai-agents-in-opensearch.md

## **Flow agent**

Flow agent is the simplest type of agent. It orchestrates multiple tools in a workflow and runs the tools sequentially, as outlined in configuration. The flow agent acts as an orchestrator of a fixed workflow, which invokes tools and coordinates the tools so that one tool’s output can become another tool’s input.

Flow agent is useful for [retrieval-augmented generation (RAG)](https://opensearch.org/blog/using-opensearch-for-retrieval-augmented-generation-rag/), where the agent invokes a tool to retrieve the knowledge base, and then invokes another tool to call the LLM while passing the knowledge base retrieved from the previous step as input, along with the query. This is ideal for implementing *one-shot prompts*.

For more information on Flow agents, read [here](https://docs.opensearch.org/latest/ml-commons-plugin/agents-tools/agents/flow/).

## **Conversational flow agent**

Much like the flow agent, Conversational flow agent runs tools sequentially, in the order specified in its configuration, to execute a fixed specified workflow.

> However, the conversational flow agent also stores conversation history in its memory index between prompts in the session. This history allows users to ask follow-up questions that the agent can address with that historical context. This makes the conversational agent a great fit for building chatbots. It is a type of a *few-shot prompts* with the conversation history as the context.

Author: @ylwu-amzn

> The conversation history is stored in an index in the OpenSearch cluster, much like any other data, which makes it conveniently accessible for querying and inspection by agents or humans, for auditing, reasoning and debugging purposes.

Author: @pyek-bot please suggest where to link it to and the text that's needed for that, as the current text doesn't directly mention the memory APIs.
Also please provide the required link

For specific questions, the agent uses the *Chain-of-Thought (CoT)* process to select the best tool from the configured tools for providing a response to the question.

Conversational agents are useful for creating a chatbot that employs RAG.

## **Plan-execute-reflect agent**

Plan-execute-reflect agent is a more elaborate form of dynamic workflow. It dynamically plans, executes, and refines multi-step workflows to solve complex tasks.

Internally, a plan-execute-reflect agent uses the multi-agent pattern in OpenSearch. Multi-agent enables chaining agents and defining sub-agents. In the case of the plan-execute-reflect agent, it employs a CoT agent for the planning phase, and a conversational agent to execute each individual step in the plan. The agent automatically selects the most appropriate tool for each step based on tool descriptions and context.

Plan-execute-reflect agent is ideal for long-running, exploratory processes, which benefit from iterative reasoning and adaptive execution, such as conducting research or performing automated root cause analysis (RCA).

> The OpenSearch project itself uses these agents for [Agentic search with natural language queries](https://github.com/opensearch-project/ml-commons/blob/main/docs/tutorials/agentic_search/agentic_search_llm_generated_type.md), and for the [release automation agent OSCAR](https://github.com/opensearch-project/oscar-ai-bot).

Reviewer: Yes, agentic search does not use PER

For more information on the plan-execute-reflect agent type, check out this [blog post](https://opensearch.org/blog/intelligent-troubleshooting-using-opensearch-3-0s-plan-execute-reflect-agent/).

## **Agentic memory**

As we saw, agents can use a session cache to store previous prompts as context for the conversation. This means that follow up questions on the same conversation can take the preceding questions and answers into account. This cache, however, is limited to the current session.

> Agentic memory, a new feature introduced in [OpenSearch 3.2](https://opensearch.org/blog/introducing-opensearch-3-2-next-generation-search-and-anayltics-with-enchanced-ai-capabilities/), enables extending beyond the boundaries of a single session, and maintaining persistent memory across sessions. This enables agents to learn from past conversations, for a deeper level of personalization and context. You can find more information about [agentic memory here](https://docs.opensearch.org/latest/ml-commons-plugin/api/agentic-memory-apis/index/).

Reviewer 2: this is a good point, we have memory api for a long time, agentic memory is an advanced new feature, but we need to explain the memory api first and explain what's new in agentic memory comparing to the existing memory. @pyek-bot

> Agentic memory, a new feature introduced in [OpenSearch 3.2](https://opensearch.org/blog/introducing-opensearch-3-2-next-generation-search-and-anayltics-with-enchanced-ai-capabilities/), enables extending beyond the boundaries of a single session, and maintaining persistent memory across sessions. This enables agents to learn from past conversations, for a deeper level of personalization and context. You can find more information about [agentic memory here](https://docs.opensearch.org/latest/ml-commons-plugin/api/agentic-memory-apis/index/).

Author: @mingshl @pyek-bot can you suggest edits to the paragraph to cover the memory API, how the new agentic memory is different, and the "significant changes" mentioned for 3.3? (as the release is coming, and we want to blog post to be current)

> Agentic memory, a new feature introduced in [OpenSearch 3.2](https://opensearch.org/blog/introducing-opensearch-3-2-next-generation-search-and-anayltics-with-enchanced-ai-capabilities/), enables extending beyond the boundaries of a single session, and maintaining persistent memory across sessions. This enables agents to learn from past conversations, for a deeper level of personalization and context. You can find more information about [agentic memory here](https://docs.opensearch.org/latest/ml-commons-plugin/api/agentic-memory-apis/index/).

Author: @mingshl @pyek-bot

> Agentic memory, a new feature introduced in [OpenSearch 3.2](https://opensearch.org/blog/introducing-opensearch-3-2-next-generation-search-and-anayltics-with-enchanced-ai-capabilities/), enables extending beyond the boundaries of a single session, and maintaining persistent memory across sessions. This enables agents to learn from past conversations, for a deeper level of personalization and context. You can find more information about [agentic memory here](https://docs.opensearch.org/latest/ml-commons-plugin/api/agentic-memory-apis/index/).

Author: @mingshl @pyek-bot reiterating my request: can you suggest edits to the paragraph to cover the memory API, how the new agentic memory is different, and the "significant changes" mentioned for 3.3?

> Agentic memory, a new feature introduced in [OpenSearch 3.2](https://opensearch.org/blog/introducing-opensearch-3-2-next-generation-search-and-anayltics-with-enchanced-ai-capabilities/), enables extending beyond the boundaries of a single session, and maintaining persistent memory across sessions. This enables agents to learn from past conversations, for a deeper level of personalization and context. You can find more information about [agentic memory here](https://docs.opensearch.org/latest/ml-commons-plugin/api/agentic-memory-apis/index/).

Author: @dhrubo-os referring to an earlier comment by @mingshl:

> this is a good point, we have memory api for a long time, agentic memory is an advanced new feature, but we need to explain the memory api first and explain what's new in agentic memory comparing to the existing memory. @pyek-bot

Can you make sure your suggested copy addresses the above?

Conversational agents are useful for creating a chatbot that employs RAG.

The OpenSearch project itself uses these agents for [Agentic search with natural language queries](https://github.com/opensearch-project/ml-commons/blob/main/docs/tutorials/agentic_search/agentic_search_llm_generated_type.md).

## **Plan-execute-reflect agent**

Plan-execute-reflect agent is a more elaborate form of dynamic workflow. It dynamically plans, executes, and refines multi-step workflows to solve complex tasks.

Internally, a plan-execute-reflect agent uses the multi-agent pattern in OpenSearch. Multi-agent enables chaining agents and defining sub-agents. In the case of the plan-execute-reflect agent, it employs a CoT agent for the planning phase, and a conversational agent to execute each individual step in the plan. The agent automatically selects the most appropriate tool for each step based on tool descriptions and context.

Plan-execute-reflect agent is ideal for long-running, exploratory processes, which benefit from iterative reasoning and adaptive execution, such as conducting research or performing automated root cause analysis (RCA).

> The OpenSearch project itself uses these agents for the [release automation agent OSCAR](https://github.com/opensearch-project/oscar-ai-bot).

Author: @ylwu-amzn who can assist in mapping out OSCAR to the correct agentic framework it leverages and demonstrates? can you tag them on this?
I think it's a powerful example and I'd hate to miss the opportunity to highlight it

> The OpenSearch project itself uses these agents for the [release automation agent OSCAR](https://github.com/opensearch-project/oscar-ai-bot).

Reviewer 2: @peterzhuamazon this is a quick question, I don't think OSCAR used opensearch agent. we can remove this line if that's the truth

---
layout: post
title: "Introduction to AI agents in OpenSearch: From simple flow agents to advanced ReAct multi-agent systems"
authors: 
    - dhorovits
    - mingshl
date: 2025-09-30
categories: 
    - technical-post
excerpt: "Explore the diverse range of AI agents in OpenSearch, from simple flow agents to advanced ReAct multi-agent systems. Learn how to choose the right agent for tasks like RAG, chatbots, research, and automated root cause analysis."
---

An _AI agent_ (or simply _agent_) is a coordinator that uses a large language model (LLM) to solve a problem. After the LLM reasons and decides what action to take, the agent coordinates the action execution.

OpenSearch includes a portfolio of agents suitable for a variety of AI tasks, such as retrieval-augmented generation (RAG), chatbots, research, and automated root cause analysis (RCA).

> In this blog, we'll explore the various agents supported by OpenSearch and provide guidance on how to choose the right agent for your task. To build a clear understanding, we'll start with the simplest agents and move on to more complex ones.

Reviewer 3: ```suggestion
In this blog post, we'll explore the various agents supported by OpenSearch and provide guidance on how to choose the right agent for your task. To build a clear understanding, we'll start with the simplest agents and move on to more complex ones.
```

## Flow agent

A _flow agent_ is the simplest type of agent. It orchestrates multiple tools in a workflow and runs them sequentially, in the order defined in the agent configuration. The flow agent acts as an orchestrator of a fixed workflow, which invokes tools and coordinates the tools so that one tool's output becomes another tool's input.

> Flow agents are useful for [retrieval-augmented generation (RAG)](https://opensearch.org/blog/using-opensearch-for-retrieval-augmented-generation-rag/). During RAG, the agent first invokes a tool to retrieve information from a knowledge base, then invokes another tool to call the LLM while passing the knowledge base retrieved from the previous step as input, along with the query. This is ideal for implementing *one-shot prompts*.

Reviewer 3: ```suggestion
Flow agents are useful for [RAG](https://opensearch.org/blog/using-opensearch-for-retrieval-augmented-generation-rag/). During RAG, the agent first invokes a tool to retrieve information from a knowledge base, then invokes another tool to call the LLM while passing the knowledge base retrieved from the previous step as input, along with the query. This is ideal for implementing *one-shot prompts*.
```

For more information, see [Flow agents](https://docs.opensearch.org/latest/ml-commons-plugin/agents-tools/agents/flow/).

## Conversational flow agent

Much like a flow agent, a _conversational flow agent_ runs tools sequentially, in the order specified in its configuration, to execute a fixed workflow.

> However, the conversational flow agent also stores conversation history in its memory index between prompts in a session. This history allows users to ask follow-up questions that the agent can address using that historical context. The conversational flow agent is ideal for a chatbot that needs to execute a predetermined, multi-step process for most of its user interactions, but also needs to remember the context of the conversation for follow-up questions. It achieves this by using a *few-shot prompt*, in which the conversation history serves as context for each new prompt. 

Reviewer 3: ```suggestion
However, the conversational flow agent also stores conversation history in its memory index between prompts in a session. This history allows users to ask follow-up questions that the agent can address using that historical context. The conversational flow agent is ideal for a chatbot that needs to execute a predetermined, multi-step process for most of its user interactions but also needs to remember the context of the conversation for follow-up questions. It achieves this by using a *few-shot prompt*, in which the conversation history serves as context for each new prompt. 
```

The conversation history is stored in an index in the OpenSearch cluster, much like any other data. It is conveniently accessible for querying and inspection by agents or humans for auditing, reasoning, and debugging. The data can be accessed using the [Memory APIs](https://docs.opensearch.org/latest/ml-commons-plugin/api/memory-apis/index/). For more information see [Agentic memory](#agentic-memory).

## Conversational agent

Unlike the previously described agents, a _conversational agent_ isn't bound by a fixed workflow. Instead, it supports dynamic workflows by using LLM-based *reasoning* to provide a response, based on the *ReAct (reasoning & acting)* agentic framework. The LLM reasons iteratively to decide what action to take until it obtains the final answer or reaches the iteration limit.

> The conversational agent bases its reasoning on two sources: the LLM’s built-in knowledge base and a set of tools that enable it to fetch additional context beyond that knowledge.

Reviewer 3: ```suggestion
The conversational agent bases its reasoning on two sources: the LLM's built-in knowledge base and a set of tools that enable it to fetch additional context beyond that knowledge.
```

For specific questions, the agent uses the *Chain-of-Thought (CoT)* process to select the best tool from the configured tools for providing a response to the question.

Similarly to conversational flow agent, conversational agents store conversation history so that users can ask follow-up questions. The workflow of a conversational agent is variable, based on follow-up questions.

Conversational agents are useful for creating chatbots that use RAG.

OpenSearch uses these agents internally for [agentic search with natural language queries](https://github.com/opensearch-project/ml-commons/blob/main/docs/tutorials/agentic_search/agentic_search_llm_generated_type.md).

## Plan-execute-reflect agent

A plan-execute-reflect agent uses a more elaborate form of dynamic workflow. It dynamically plans, executes, and refines multi-step workflows to solve complex tasks.

> Internally, a plan-execute-reflect agent uses the multi-agent pattern in OpenSearch, which enables chaining agents and defining subagents. The plan-execute-reflect agent uses a CoT agent for the planning phase, and a conversational agent to execute each individual step in the plan. The agent automatically selects the most appropriate tool for each step based on tool descriptions and context.

Reviewer 3: ```suggestion
Internally, a plan-execute-reflect agent uses the multi-agent pattern in OpenSearch, which enables chaining agents and defining subagents. The plan-execute-reflect agent uses a CoT agent for the planning phase and a conversational agent to execute each individual step in the plan. The agent automatically selects the most appropriate tool for each step based on tool descriptions and context.
```

The plan-execute-reflect agent is ideal for long-running, exploratory processes, which benefit from iterative reasoning and adaptive execution, such as conducting research or performing automated RCA. 

For more information about the plan-execute-reflect agent, check out this [blog post](https://opensearch.org/blog/intelligent-troubleshooting-using-opensearch-3-0s-plan-execute-reflect-agent/).

## Agent tools and external data integration

> Agents are only as effective as the portfolio of toolsat their disposal. OpenSearch's agentic framework enables agents to invoke both built-in and external tools using standardized protocols.

Reviewer 3: ```suggestion
Agents are only as effective as the portfolio of tools at their disposal. OpenSearch's agentic framework enables agents to invoke both built-in and external tools using standardized protocols.
```

### Built-in agent tools

OpenSearch provides a comprehensive suite of prebuilt [tools](https://docs.opensearch.org/latest/ml-commons-plugin/agents-tools/tools/index/) that can be seamlessly configured with the different agent types. These native tools enable agents to efficiently access and manipulate search data within OpenSearch, providing essential capabilities for data retrieval, analysis, and management operations.
 
### External data source integration using MCP connectors

Model Context Protocol (MCP) has been gaining popularity as a communication standard for agentic workflows, simplifying the way AI agents interact with external tools.

> OpenSearch extends agent capabilities through [MCP Connectors](https://docs.opensearch.org/latest/ml-commons-plugin/agents-tools/mcp/index/), enabling integration with external MCP servers and their associated tools. This powerful feature is supported by plan-execute-reflect agents and conversational agents, allowing them to use external data sources and services, and significantly expanding their analytical and operational capabilities beyond the confines of OpenSearch itself. The connectors support the Streamable HTTP and the Server-Sent Events (SSE) protocols, providing flexibility in connecting to various external MCP servers. These agents can seamlessly combine internal OpenSearch tools with external MCP server tools, creating unified workflows that enable sophisticated multi-source data analysis and decision-making processes across diverse data sources and services.

Reviewer 3: ```suggestion
OpenSearch extends agent capabilities through [MCP connectors](https://docs.opensearch.org/latest/ml-commons-plugin/agents-tools/mcp/index/), enabling integration with external MCP servers and their associated tools. This powerful feature is supported by plan-execute-reflect agents and conversational agents, allowing them to use external data sources and services and significantly expanding their analytical and operational capabilities beyond the confines of OpenSearch itself. The connectors support the Streamable HTTP and Server-Sent Events (SSE) protocols, providing flexibility in connecting to various external MCP servers. These agents can seamlessly combine internal OpenSearch tools with external MCP server tools, creating unified workflows that enable sophisticated multi-source data analysis and decision-making processes across diverse data sources and services.
```

By providing agents with persistent, searchable memory, this feature transforms static AI interactions into dynamic, context-aware experiences that improve over time, enabling more personalized and intelligent responses while maintaining control over memory organization and retention policies. For more information, see [Agentic memory](https://docs.opensearch.org/latest/ml-commons-plugin/agentic-memory/).

## An agentic experience with an MCP server

External agents can also benefit from accessing OpenSearch capabilities exposed as tools over standard MCP. OpenSearch offers two comprehensive [MCP server solutions](https://opensearch.org/blog/introducing-mcp-in-opensearch/) to meet diverse integration needs:
 

> 1. **Built-in MCP Server**: The [built-in MCP server](https://docs.opensearch.org/latest/ml-commons-plugin/api/mcp-server-apis/index/) provides native integration within OpenSearch, exposing a streamable HTTP MCP API that enables dynamic tool registration and removal. This solution offers seamless access to core OpenSearch capabilities, allowing agents to invoke powerful [tools](https://docs.opensearch.org/latest/ml-commons-plugin/agents-tools/tools/index/) for real-time data queries and analysis. Since it makes direct API calls to OpenSearch through `/_plugins/_ml/mcp`, it natively supports Role-Based Access Control (RBAC), ensuring secure and granular access management.

Reviewer 3: ```suggestion
1. **Built-in MCP server**: The [built-in MCP server](https://docs.opensearch.org/latest/ml-commons-plugin/api/mcp-server-apis/index/) provides native integration within OpenSearch, exposing a Streamable HTTP MCP API that enables dynamic tool registration and removal. This solution offers seamless access to core OpenSearch capabilities, allowing agents to invoke powerful [tools](https://docs.opensearch.org/latest/ml-commons-plugin/agents-tools/tools/index/) for real-time data queries and analysis. Since it makes direct API calls to OpenSearch through `/_plugins/_ml/mcp`, it natively supports Role-Based Access Control (RBAC), ensuring secure and granular access management.
```

> 2. **Standalone MCP Server**: The [standalone MCP server](https://github.com/opensearch-project/opensearch-mcp-server-py/blob/main/USER_GUIDE.md) delivers flexible deployment options through the `opensearch-mcp-server-py` PyPI package. This solution supports multiple protocols including `stdio`, Streamable HTTP, and Server-Sent Events (SSE), making it compatible with various AI agent frameworks. Advanced features include tool filtering capabilities and multi-cluster connectivity, enabling sophisticated agentic workflows across distributed environments.

Reviewer 3: ```suggestion
2. **Standalone MCP server**: The [standalone MCP server](https://github.com/opensearch-project/opensearch-mcp-server-py/blob/main/USER_GUIDE.md) delivers flexible deployment options through the `opensearch-mcp-server-py` PyPI package. This solution supports multiple protocols, including `stdio`, Streamable HTTP, and SSE, making it compatible with various AI agent frameworks. Advanced features include tool filtering capabilities and multi-cluster connectivity, enabling sophisticated agentic workflows across distributed environments.
```
